x/tokenfactory/migrations/v2/store.go

```diff
@@ -0,0 +1,112 @@
+
+	"github.com/neutron-org/neutron/v4/x/tokenfactory/types"
+	v1beta1types "github.com/neutron-org/neutron/v4/x/tokenfactory/types/v1beta1"
+)
+
+type TokenFactoryKeeper interface {
+	AssertIsHookWhitelisted(ctx sdk.Context, denom string, contractAddress sdk.AccAddress) error
+}
+
+// MigrateStore performs in-place store migrations.
+// The migration adds the new tokenfactory params WhitelistedHooks
+func MigrateStore(ctx sdk.Context, cdc codec.BinaryCodec, storeKey storetypes.StoreKey, keeper TokenFactoryKeeper) error {
+	// NOTE: this must happen first since the migrateHooks relies on the new params
+	if err := migrateParams(ctx, cdc, storeKey); err != nil {
+		return err
+	}
+	if err := migrateHooks(ctx, storeKey, keeper); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var WhitelistedHooks = []*types.WhitelistedHook{
+	{ // xASTRO balances tracker
+		CodeID:       944,
+		DenomCreator: "neutron1zlf3hutsa4qnmue53lz2tfxrutp8y2e3rj4nkghg3rupgl4mqy8s5jgxsn",
+	},
+	{ // NFA.zoneV1
+		CodeID:       1265,
+		DenomCreator: "neutron1pwjn3tsumm3j7v7clzqhjsaukv4tdjlclhdytawhet68fwlz84fqcrdyf5",
+	},
+}
+
+func migrateParams(ctx sdk.Context, cdc codec.BinaryCodec, storeKey storetypes.StoreKey) error {
+	ctx.Logger().Info("Migrating tokenfactory params...")
+
+	// fetch old params
+	store := ctx.KVStore(storeKey)
+	bz := store.Get(types.ParamsKey)
+	if bz == nil {
+		return errors.New("cannot fetch tokenfactory params from KV store")
+	}
+	var oldParams v1beta1types.Params
+	cdc.MustUnmarshal(bz, &oldParams)
+
+	// add new param values
+	newParams := types.Params{
+		DenomCreationFee:        oldParams.DenomCreationFee,
+		DenomCreationGasConsume: oldParams.DenomCreationGasConsume,
+		FeeCollectorAddress:     oldParams.FeeCollectorAddress,
+		WhitelistedHooks:        WhitelistedHooks,
+	}
+
+	// set params
+	bz, err := cdc.Marshal(&newParams)
+	if err != nil {
+		return err
+	}
+	store.Set(types.ParamsKey, bz)
+
+	ctx.Logger().Info("Finished migrating tokenfactory params")
+
+	return nil
+}
+
+func migrateHooks(ctx sdk.Context, storeKey storetypes.StoreKey, keeper TokenFactoryKeeper) error {
+	ctx.Logger().Info("Migrating tokenfactory hooks...")
+
+	// get hook store
+	store := prefix.NewStore(ctx.KVStore(storeKey), []byte(types.DenomsPrefixKey))
+	iterator := storetypes.KVStorePrefixIterator(store, []byte{})
+
+	for ; iterator.Valid(); iterator.Next() {
+		keyParts := strings.Split(string(iterator.Key()), types.KeySeparator)
+
+		// Hooks and authorityMetadata are in the same store, we only care about the hooks
+		if keyParts[2] == types.BeforeSendHookAddressPrefixKey {
+			denom := keyParts[1]
+			contractAddr, err := sdk.AccAddressFromBech32(string(iterator.Value()))
+			if err != nil {
+				return errors.New("cannot parse hook contract address")
+			}
+
+			err = keeper.AssertIsHookWhitelisted(ctx, denom, contractAddr)
+			if err != nil {
+				// If hook is not whitelisted delete it
+				store.Delete(iterator.Key())
+			}
+		}
+	}
```

Comment:
Reposting from slack for others context:

So thinking about it more I'm not sure I totally agree with this. After the migration when we use this check in MsgServer it doesn't really matter why the hook isn't whitelisted (invalid contract, invalid denom, no whitelist value) we are always going to return an error. So we might as well use AssertIsHookWhitelisted to generate the errors so they are as informative as possible. If it returns false, nil for the case where there are no "errors" but the hook is not whitelisted then the message server  code has to be
	```
whitelisted, err := server.Keeper.AssertIsHookWhitelisted(ctx, msg.Denom, cwAddr)
		if err != nil {
			return nil, err
		}
		if !whitelisted {
			return nil, types.ErrBeforeSendHookNotWhitelisted
		}
```
Which feels messy. Also, we no longer have the context to return an informative error message like we do here:  https://github.com/neutron-org/neutron/blob/a6eae66545df6e128da2c7cac81ff4c1c02deb7b/x/tokenfactory/keeper/params.go#L51
In the case of the migration, I agree that it's not quite perfect. but given that we are migrating a small set of specific hooks I think we can fairly assume that those are correct. If there are other hooks that throw an error as part of the check we do not want to stop the migration. For example, if someone added a hook before the upgrade the pointed to a non-existent contract we wouldn't want the migration to fail.
Let me know what you think?


I will deal with the array case better.

Reply:
okay, I see. what do you think about the following approach then?
1. make the errors returned from `AssertIsHookWhitelisted` more specific in terms of types:
<img width="1420" alt="image" src="https://github.com/neutron-org/neutron/assets/34917380/ee946213-2872-4c54-ae67-8a6360dd755c">


2. remove hooks only in case of the `types.ErrBeforeSendHookNotWhitelisted error`:
<img width="1426" alt="image" src="https://github.com/neutron-org/neutron/assets/34917380/e47d04e8-7d80-4ee9-8d28-b49e8a95a527">

I think it's a quite good enough alternative here